### PR TITLE
Remove leftover Discord implementation

### LINKED ISSUE #1370: Our Discord integration violates the GPL
**Describe the bug**

We cannot link with Discord's Game SDK as the Game SDK is not available under a free license.

**Describe the solution you'd like**

Use the legacy [`discord-rpc`](https://github.com/discord/discord-rpc) library instead, which _is_ available under a free license (MIT).

This issue also pauses any _mass rollouts_ of binaries until it has been resolved.

**Describe alternatives you've considered**

Contact all historical contributors and change our license. Doing this is not practical at this point in time—it will take a long time, and it's faster to just change our code.

**Version**

245551c

**Additional context**

Discord announcement of the Game SDK with extensive bikeshedding on this matter: https://github.com/discord/discord-rpc/issues/290#issuecomment-488151732

Relevant conversation in #mta.dev: https://discordapp.com/channels/278474088903606273/366384007535001612/697937233151524884

### PR DESCRIPTION
**Pull request by Lpsd**
Fixes #1370

Removes the leftover Discord implementation - aside from CServerInfo / CQueryReceiver stuff introduced in the original PR, which is actually a benefit.

### CODE REVIEW

File: Shared/sdk/net/Packets.h
Comment (TheNormalnij) [inline Shared/sdk/net/Packets.h]: I think this causes desync problem.  Will we merge this in 1.6?

Comment (Lpsd) [reply Shared/sdk/net/Packets.h]: We could keep the packet declared here but not use it, I think that's fine

**Comment by AlexTMjugador** [comment]
#1570 was opened to bring back the Discord integration, but I take @0x416c69 doesn't seem that active these days?

By the looks of it, I think that using the deprecated library instead of the latest, shiniest SDK can be a viable option. Discord would make even more developers unhappy if they made the old library stop working, as it is presumably being used in some big proprietary games: https://github.com/discord/discord-rpc/pull/306#issuecomment-696398048. Hopefully, those big proprietary games are big and proprietary enough to make a migration to the new SDK a no-go, so I'd say Discord is compelled to align with the interests of GPL projects to maintain the status quo.

**Comment by Lpsd** [comment]
Not sure if they're still around or have plans to work on it - but we shouldn't have unused code sitting around.

@Deihim007 mentioned implementing it themselves, differently than #1570 (but explicitly no promises 😅)

**Comment by AlexTMjugador** [comment]
Yeah, I agree with removing this old code if nobody is interested in maintaining it properly (and maybe even maintaining a `discord-rpc` fork if the need arises). Technical debt can only increase by keeping it around for no reason. Still, I personally find the current situation with Discord rich presence integration libraries quite unsatisfactory.

**Comment by 0x416c69** [comment]
> I take @0x416c69 doesn't seem that active these days?

I do not have an interest in doing the clean up for the git commits and at that time I got very busy to do so. My PR was complete and flawless and ready to be merged.

> By the looks of it, I think that using the deprecated library instead of the latest, shiniest SDK can be a viable option.

Originally I've used the newer SDK and actually I had a complete working version of it (it's still in the PRs somewhere) but because of how unstable it was and it caused so many crashes (As reported by @Dutchman101 ) and also the license issue (MTASA license does not allow to make use of it AFAIK) I had to refactor and clean up the old SDK and implement the rich presence over it which was decided and done as planned. To this day so many people are still using the old library and as I stated earlier the repository I've made for it is gonna be maintained and if the protocol of the IPC is changed I'm gonna implement it and make it work again but discord will be unlikely to change the protocol as it is unnecessary and would break compatibility.

I still don't understand why has no developer taken interest in my PR despite so many people wanting this feature on the MTASA community. Anyways as I said my PR is complete and I no longer wish to contribute further, close it if you wish.

**Comment by Dutchman101** [comment]
@0x416c69
> I still don't understand why has no developer taken interest in my PR despite so many people wanting this feature on the MTASA community. Anyways as I said my PR is complete and I no longer wish to contribute further, close it if you wish.

Last week we discussed this (PR #1570) internally, to figure out why it stalled and got no code review attention.. it was kinda sad, because of the hype surrounding Discord Rich integration within MTA (there was a lot of developer attention, which suddenly dropped when you had to redo it). However, as a result of the discussion, maybe it will get some eyes soon.. so let's hold up before closing it

If it turns out to be well-designed and not really introduce bugs or more pending work, who knows it can do without your activity, or that another dev can easily jump in for small patches to anything that may arise during code review or right after implementation. Although that being the author is obviously better, we'll see about it?

**Comment by Lpsd** [comment]
Firstly, thank you for your contribution on #1570 

In all honesty there's just not that many people around here confident enough, or with the knowledge, to test and review your PR. For those that can, it then boils down to whether they are interested in it or not (luck of the draw). Unfortunately lots of other great PRs have collected dust this way, it isn't exclusive to #1570

Regardless, I think this cleanup is necessary in master before reviewing #1570 (will then need rebasing or whatever). 

I'd be happy to look into your PR, however if you're not around to contribute further do we have permission to use it entirely (i.e fork and continue changes ourselves)?

**Comment by 0x416c69** [comment]
> to figure out why it stalled and got no code review attention

I was in contact with @qaisjp on discord and he instructed me what to do as he told on my PR but then I got busy and I couldn't get back to this so it could be part of the reason why.

> or that another dev can easily jump in for small patches to anything that may arise during code review or right after implementation

Hoping to see that one day.
I just no longer have an interest in doing it. I'm more interested in modding and expanding the game itself now and I will hopefully make some PRs on it someday.

> For those that can, it then boils down to whether they are interested in it or not (luck of the draw). Unfortunately lots of other great PRs have collected dust this way,

Of course, this is the case for many non-profit open source projects, I understand. But so many people and devs were really interested at the time but I guess they lost interest just as I did.

> if you're not around to contribute further do we have permission to use it entirely (i.e fork and continue changes ourselves)?

Sure, go ahead.

**Comment by Lpsd** [comment]
I've reverted the enum removals in latest commit (they would cause desync). This is ready to review.

File: Client/mods/deathmatch/logic/CClientGame.cpp
Comment (TheNormalnij) [inline Client/mods/deathmatch/logic/CClientGame.cpp:627]: Is it safe, when client and server have different versions?

Comment (Lpsd) [reply Client/mods/deathmatch/logic/CClientGame.cpp:627]: I just checked. We should bump the bitstream in this PR (something like `Discord_Cleanup`). 

If the server has `Discord_InitialImplementation` but not `Discord_Cleanup`, then the client should send an empty string (otherwise player join will fail).

If server has `Discord_Cleanup` then we don't need to send anything.

As for the server -> client bitstream, that doesn't seem to matter. 

For the Lua definition, `setPlayerDiscordJoinParams`, the server requests an RPC - `SET_DISCORD_JOIN_PARAMETERS` - which won't exist if the client is newer than server (no case = ignored).

It also tries to send a `CServerInfoSyncPacket` (`PACKET_ID_SERVER_INFO_SYNC`) on player join with flag `SERVER_INFO_FLAG_ALL`, but there isn't a case for handling that flag in `CPacketHandler::Packet_ServerInfoSync`. So if the client is newer, the packet is ignored, and if the client is outdated nothing happens currently anyway.

**Comment by Lpsd** [comment]
I've made the bitstream changes - ready to review x3

**Review by patrikjuvonen — APPROVED** [review]
Tested lightly, works. I scrolled through the changes.

File: Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp
Comment (botder) [inline Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp:64]: Please do not remove the Lua functions. Keep a dummy function and deprecate it.

File: Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp
Comment (botder) [inline Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp:144]: Please do not remove the Lua functions. Keep a dummy function and deprecate it.

File: Server/mods/deathmatch/logic/packets/CPlayerJoinDataPacket.cpp
Comment (botder) [inline Server/mods/deathmatch/logic/packets/CPlayerJoinDataPacket.cpp:39]: Please do not _silently_ hide this part of the old BitStream version, just because it was at the end of it.
Read the string, but only if the BitStream version **v** is `v >=Discord_InitialImplementation && v < Discord_Cleanup` , discard the value and add a comment what and why happened.

File: Client/mods/deathmatch/logic/CClientGame.cpp
Comment (botder) [inline Client/mods/deathmatch/logic/CClientGame.cpp:630]: Just for confirmation, does this send the same bits like the old code?

**Review by botder — CHANGES_REQUESTED** [review]

File: Server/mods/deathmatch/logic/packets/CPlayerJoinDataPacket.cpp
Comment (Pirulax) [reply Server/mods/deathmatch/logic/packets/CPlayerJoinDataPacket.cpp:39]: what's the point of reading it from the bitstream if the read value is going to be discarded?

File: Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp
Comment (Pirulax) [reply Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp:64]: Tip: A lambda can be used.

**Comment by Pirulax** [comment]
LGTM.
Do the requested changes from Botder's review, and once he gives the green light it can be merged IMHO.
Worst case scenario we'll pull it back from the nightly.

File: Server/mods/deathmatch/logic/packets/CPlayerJoinDataPacket.cpp
Comment (botder) [reply Server/mods/deathmatch/logic/packets/CPlayerJoinDataPacket.cpp:39]: Nevermind, the BitStream version will be different, if that packet ever changes in the future.

**Comment by Lpsd** [comment]
Requested changes have been made - ready for final review

**Review by botder — APPROVED** [review]

**Pull request merged**
